pkg/memory-monitor/src/monitor/main.c

```diff
@@ -0,0 +1,203 @@
+            pthread_cancel(resources_to_cleanup.threads_to_finish.threads[i]);
+        }
+        free(resources_to_cleanup.threads_to_finish.threads);
+        resources_to_cleanup.threads_to_finish.threads = NULL;
+        resources_to_cleanup.threads_to_finish.count = 0;
+    }
+    // Close the FDs
+    if (resources_to_cleanup.fds_to_close.fds != NULL) {
+        for (size_t i = 0; i < resources_to_cleanup.fds_to_close.count; i++) {
+            if (resources_to_cleanup.fds_to_close.fds[i] != -1) {
+                close(resources_to_cleanup.fds_to_close.fds[i]);
+            }
+        }
+        free(resources_to_cleanup.fds_to_close.fds);
+        resources_to_cleanup.fds_to_close.fds = NULL;
+        resources_to_cleanup.fds_to_close.count = 0;
+    }
+}
+
+
+void onexit_cleanup() {
+    // Stop the threads and close the FDs
+    onreload_cleanup();
+    // Destroy the semaphore
+    if (semaphores_initialized) {
+        sem_destroy(&reload_semaphore);
+    }
+    // Close the system log
+    if (syslog_opened) {
+        syslog(LOG_INFO, "Stopping\n");
+        closelog();
+    }
+}
+
+void term_handler(int signo) {
+    exit(signo); // This will cause the onexit_cleanup func to be called, closing the system log and cleaning up the resources
+}
+
+void hup_handler(int signo) {
+    (void)signo; // Unused
+    sem_post(&reload_semaphore);
+    // Stop the thread
+    onreload_cleanup();
+}
+
+int main(int argc, char *argv[]) {
+    pid_t pid, sid;
+    bool daemonize = true;
+    int opt;
+
+    while ((opt = getopt(argc, argv, "f")) != -1) {
+        switch (opt) {
+            case 'f':
+                daemonize = false;
+                break;
+            default:
+                exit(EXIT_FAILURE);
+        }
+    }
+
+    if (daemonize) {
+        // Fork off the parent process
+        pid = fork();
+        if (pid < 0) {
+            exit(EXIT_FAILURE);
+        }
+        // If we got a good PID, then we can exit the parent process
+        if (pid > 0) {
+            exit(EXIT_SUCCESS);
+        }
+
+        // Change the file mode mask
+        umask(0);
+
+        // Create a new SID for the child process
+        sid = setsid();
+        if (sid < 0) {
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    // Move the process to the root cgroup
+    cgroup_move_process_to_root_memory(getpid());
+
+    // Save the binary location, as the handler script is in the same directory
+    // First, get the full path to the binary
+    char binary_full_path[PATH_MAX + 1];
+    if (realpath(argv[0], binary_full_path) == NULL) {
+        exit(EXIT_FAILURE);
+    }
+    //Then - copy the directory part into the global variable
+    strncpy(binary_location_g, dirname(binary_full_path), sizeof(binary_location_g) - 1);
+
+    // Create a new application directory, if it doesn't exist
+    if (access(APP_DIR, F_OK) == -1) {
+        if (mkdir(APP_DIR, 0755) == -1) {
```

Comment:
`group` and `other` also need execute perm?

Reply:
I'll try to remove it now.

Reply:
Mmm... actually, it's a directory permission. Which means X here is for accessing/listing. And that's expected from a directory with the output and configuration that a user can access it via SSH for example. I can try to remove the W permission from the group and others, but I think X should present.

Reply:
my bad, you are right, keep the `x`.